### ISSUE TITLE
Flexible Field-Replaceable Unit (FRU) or Equipment Modules

### ISSUE THREAD

**Issue by mokun** [mars-sim/mars-sim#860]
**Describe the Feature**
- What is the best architecture for modeling how to swap in and out *field-replaceable unit* or equipment modules in a vehicle ?
- See [this](https://en.wikipedia.org/wiki/Field-replaceable_unit) or [this](https://en.wikipedia.org/wiki/Line-replaceable_unit) on what FRU or LRU are all about.

**Background**
- Currently, `GenericContainer` extends `Equipment` and implements `Container` interface.
- Other than `GenericContainer`, we have an `BuildingKit` class that we never put into use.
- We have recently created `Battery` class to model what a battery module would behave in a vehicle in tracking the energy usage.

**Discuss potential strategies**
- Q1 : Should *Battery Module* and *Fuel Cell Stack* fall under `Equipment` class ?
- Q2 : Should we create `ModularUnit` and make it extends `Equipment` ?
- In future, one should be able to swap things out in within building or a weather station out there as well

**Comment by mokun** [comment]
Well, that's not a whole lot of EVA suits for now. A 24 person settlement may be 48 suits only if we keep a spare/backup suit for everyone.

**Comment by mokun** [comment]
Currently, a battery module doesn't need a pulse yet since we haven't consider any parasitic drain on the battery yet nor the natural discipation of energy.

Also, we can make use of the ,astwr-slave relationship. If the battery a "slave" to the vehicle. then whenever the vehicle receives a pulse, it can propagate the

**Comment by bevans2000** [comment]
Ideally the FRU should be a  Equipment. This is how I see the modelling:

- Parts - stateless entities not individually modelled
- Equipment - stateful entities modelled individually but always owned, not standalone. Are not affected by time did fly, I.e. do night need a clock pulse
- Units - stateful entities directly affected by time, I.e. clock pulse

The current implementation does not matter this. Equipment take a pulse which means all the Containers are consuming processing cycles every pulse for no reason. I would like to drop Unit from Equipment and create a new Unit class to hold EVASuit, e.g Machine. This would reduce the resource consumption if the 100s of Container objects.

**Comment by mokun** [comment]
> Are not affected by time did fly, I.e. do not need a clock pulse

Yes it's a good strategy to have certain equipment  to be checked only once a sol. Don't need to send pulse once every frame.

I can see that stationary equipment such as battery, machine, server, fuel cell stack, etc should reasonably depreciate over time if being put into active use. 

A pulse that updates the condition once a sol is very reasonable in that case

**Comment by mokun** [comment]
> all the Containers are consuming processing cycles every pulse for no reason. I would like to drop Unit from Equipment and create a new Unit class to hold EVASuit, e.g Machine. This would reduce the resource consumption if the 100s of Container objects.

Should we simply separate all containers from Equipment ? 

Containers are kind of passive and we don't see the value of tracking their conditions in every frame.

On the other hand, equipment such as robot, vehicle, battery, fuel cell stack, server node, etc are mission essential. They do deserve closer monitoring and higher fidelity of modeling.

**Comment by mokun** [comment]
> Container do not really need to be Units as they are passive but they do need to be Equipment so they can be owned.

So I just reread this that you wrote earlier. 

Yes it makes sense on how you see the relationship should be. 

So would it help if we create a ContainerOwner class solely for making a container ownable by a unit ? 

I imagine this ContainerOwner would be very similar to EquipmentOwner.

This way, a container can be owned by an unit but don't have to inherit from Equipment ?!

**Comment by bevans2000** [comment]
Having a ContainerOwner is going to complicate the modelling more. I think we should retain EquipmentOwner but drop Unit from the Equipment class. This would also drop Equipment from UnitManager.  This would work fine I think.
The problem is what to do about EVASuit. Nevertheless, this needs careful thought before jumping in with changes.

**Comment by mokun** [comment]
So after we drop Equipment from Unit, may be we could experiment with ways to send pulse to Equipment on-demand !?

So right now, how often does EVA suit need the time pulse and for what purpose  ?
